rfcs/005-all-docs-index.md

```diff
@@ -0,0 +1,207 @@
+make sense to have both.
+
+The current implementation of `sizes.external` does *not* measure the length of
+a JSON representation of the data, but rather the size of an uncompressed Erlang
+term representation of the JSON. This is a somewhat awkward choice as the
+internal Erlang term representation is liable to change over time (e.g. with the
+introduction of Maps in newer Erlang releases, or plausibly even a JSON decoder
+that directly emits the format defined in the document storage RFC).
+
+Assuming we can agree on a set of sizes and how they should be calculated, the
+implementation will require two pieces: a single key for each size, mutated by
+atomic operations, and a record of the size of each revision in the ?REVISIONS
+subpsace so that a transaction can compute the delta for each document.
+
+### Clustering
+
+The `r`, `w`, `q`, and `n` values in the `cluster` object were introduced in
+CouchDB 2.x to describe the topology of a database and the default quorum
+settings for operations against it. If we wanted to bring these forward, here's
+how they'd be defined:
+
+* `r`: always fixed at 1
+
+* `w`: interpreted as the number of transaction logs that record a commit, this
+  is dependent on the `redundancy mode` for the underlying FoundationDB database
+
+* `n`: interpreted as number of storage servers that host a key, this is also
+  dependent on the `redundancy mode` for the underlying FoundationDB database
+
+* `q`: the closest analogue here would be to use the `get_boundary_keys` API and
+  report number of distinct ranges implied by the boundary keys
+
+This interpretation could lead to some surprises, though. For example, "r=1,
+w=4, n=3" is a popular configuration, but this is nonsensical for someone
+expecting to see Dynamo-style numbers. Ignoring backwards compatibility, the
+sensible thing is to point users toward the actual FoundationDB configuration
+information, and to deprecate this entire `cluster` object. Open for discussion.
+
+# Advantages and Disadvantages
+
+[NOTE]: # ( Briefly, list the benefits and drawbacks that would be realized should )
+[NOTE]: # ( the proposal be accepted for inclusion into Apache CouchDB. )
+
+# Key Changes
+
+The underlying transaction in FoundationDB must complete within 5 seconds, which
+implicitly limits the number of results that can be returned in a single
+_all_docs invocation.
+
+## Applications and Modules affected
+
+TBD depending on exact code layout going forward.
+
+## HTTP API additions
+
+None.
+
+## HTTP API deprecations
+
+The `total_rows` and `offset` fields are removed from the response to
+`_all_docs`, which now has the simpler form
+
+    {"rows": [
+        {"id":"foo", "key":"foo", "value":{"rev":"1-deadbeef..."}},
+        ...
+    ]}
+
+The following fields are removed in the dbinfo response:
+
+* `compact_running`
+
+* `disk_format_version`: this is a tricky one. We define "format versions" for
```

Comment:
Just a thought, but would choosing smart values for format versions and ORing them into a `chmod`-style bitfield work?

If that’s impractical, do we expect format definitions to mix-and-match over time, or could we define “the set of per-let format versions XYZ is 1, XZZ is 2” and so on. This would be close to what we have now where we make arbitrary changes to the db header record and each change gets an inc by one.

Reply:
I think this would devolve to "CouchDB version 4.x.y writes data using on-disk format version 12"; i.e., it'd be a property of the running CouchDB code rather than a property of a database. Individual databases are just portions of a global key space, and we don't have any background process that's sitting down and rewriting an entire database into the latest format, so at first glance we'll need to support full schema evolution for each data type we're storing in FDB.